dbms/src/Functions/runningAccumulate.cpp

```diff
@@ -70,33 +70,71 @@ class FunctionRunningAccumulate : public IFunction
 
     void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result, size_t /*input_rows_count*/) override
     {
+        size_t number_of_arguments = arguments.size();
+
+        if (number_of_arguments == 0)
+            throw Exception("Incorrect number of arguments of function " + getName() + ". Must be 1 or 2.",
+                ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+
         const ColumnAggregateFunction * column_with_states
             = typeid_cast<const ColumnAggregateFunction *>(&*block.getByPosition(arguments.at(0)).column);
+
         if (!column_with_states)
             throw Exception("Illegal column " + block.getByPosition(arguments.at(0)).column->getName()
                     + " of first argument of function "
                     + getName(),
                 ErrorCodes::ILLEGAL_COLUMN);
 
+        ColumnPtr column_with_groups;
+
+        if (number_of_arguments == 2)
+            column_with_groups = block.getByPosition(arguments[1]).column;
+
         AggregateFunctionPtr aggregate_function_ptr = column_with_states->getAggregateFunction();
         const IAggregateFunction & agg_func = *aggregate_function_ptr;
 
         AlignedBuffer place(agg_func.sizeOfData(), agg_func.alignOfData());
-        agg_func.create(place.data());
-        SCOPE_EXIT(agg_func.destroy(place.data()));
 
+        /// Will pass empty arena if agg_func does not allocate memory in arena
         std::unique_ptr<Arena> arena = agg_func.allocatesMemoryInArena() ? std::make_unique<Arena>() : nullptr;
 
         auto result_column_ptr = agg_func.getReturnType()->createColumn();
         IColumn & result_column = *result_column_ptr;
         result_column.reserve(column_with_states->size());
 
         const auto & states = column_with_states->getData();
+
+        size_t i = 0;
+
+        SCOPE_EXIT({
+            if (i > 0)
+                agg_func.destroy(place.data());
+        });
+
         for (const auto & state_to_add : states)
         {
-            /// Will pass empty arena if agg_func does not allocate memory in arena
+            if (i == 0 || (column_with_groups && column_with_groups->compareAt(i, i - 1, *column_with_groups, 1) != 0))
+            {
+                if (i > 0)
+                    agg_func.destroy(place.data());
+
+                try
+                {
+                    agg_func.create(place.data());
```

Comment:
No... if create has thrown an exception, the data is not created and calling "destroy" is incorrect.

Reply:
I take an example from [FunctionArrayReduce](https://github.com/ClickHouse/ClickHouse/blob/cc735b4554e9d2fea83d6fbcde9607c07ef5cc84/dbms/src/Functions/array/arrayReduce.cpp#L171).
 
You mean that if create fails - there's no reason to destroy?

Reply:
Fixed solution below

Reply:
> I take an example from FunctionArrayReduce.

That's strange. I think it's bug - will validate.

Reply:
Yes, the code was incorrect, see https://github.com/ClickHouse/ClickHouse/pull/8446

Reply:
That’s all?